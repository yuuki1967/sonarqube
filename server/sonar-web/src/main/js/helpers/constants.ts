/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as theme from '../app/theme';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT'
];
export const RULE_TYPES: T.RuleType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT',
  'UNKNOWN'
];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  theme.green,
  theme.lightGreen,
  theme.yellow,
  theme.orange,
  theme.red
];
export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  theme.red,
  theme.orange,
  theme.yellow,
  theme.lightGreen,
  theme.green
];

export const RATING_COLORS = [theme.green, theme.lightGreen, theme.yellow, theme.orange, '#e00'];
