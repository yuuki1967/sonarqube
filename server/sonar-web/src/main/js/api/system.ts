/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

export type SysValue = boolean | string | number | HealthType | SysValueObject | SysValueArray;
export interface SysValueObject {
  [key: string]: SysValue;
}
export interface SysValueArray extends Array<SysValue> {}

export interface SysInfoSection {
  [sectionName: string]: SysValueObject;
}

export enum HealthType {
  RED = 'RED',
  YELLOW = 'YELLOW',
  GREEN = 'GREEN'
}

export interface NodeInfo extends SysValueObject {
  'Compute Engine Logging': { 'Logs Level': string };
  Health: HealthType;
  'Health Causes': string[];
  Name: string;
  'Web Logging': { 'Logs Level': string };
}

export interface SysInfo extends SysValueObject {
  Health: HealthType;
  'Health Causes': string[];
  System: {
    'High Availability': boolean;
    'Logs Level': string;
    'Server ID': string;
  };
}

export interface ClusterSysInfo extends SysInfo {
  'Application Nodes': NodeInfo[];
  'Search Nodes': NodeInfo[];
}

export interface SystemUpgrade {
  version: string;
  description: string;
  releaseDate: string;
  changeLogUrl: string;
  downloadUrl: string;
  plugins: any;
}

export function setLogLevel(level: string): Promise<void | Response> {
  return post('/api/system/change_log_level', { level }).catch(throwGlobalError);
}

export function getSystemInfo(): Promise<SysInfo> {
  return getJSON('/api/system/info').catch(throwGlobalError);
}

export function getSystemStatus(): Promise<{ id: string; version: string; status: string }> {
  return getJSON('/api/system/status');
}

export function getSystemUpgrades(): Promise<{
  upgrades: SystemUpgrade[];
  updateCenterRefresh: string;
}> {
  return getJSON('/api/system/upgrades');
}

export function getMigrationStatus(): Promise<{
  message?: string;
  startedAt?: string;
  state: string;
}> {
  return getJSON('/api/system/db_migration_status');
}

export function migrateDatabase(): Promise<{
  message?: string;
  startedAt?: string;
  state: string;
}> {
  return postJSON('/api/system/migrate_db');
}

export function restart(): Promise<void | Response> {
  return post('/api/system/restart').catch(throwGlobalError);
}

const POLLING_INTERVAL = 2000;

function pollStatus(cb: Function): void {
  setTimeout(() => {
    getSystemStatus()
      .then(r => {
        if (r.status === 'UP') {
          cb();
        } else {
          pollStatus(cb);
        }
      })
      .catch(() => pollStatus(cb));
  }, POLLING_INTERVAL);
}

function promiseStatus(): Promise<any> {
  return new Promise(resolve => pollStatus(resolve));
}

export function restartAndWait(): Promise<any> {
  return restart().then(promiseStatus);
}
