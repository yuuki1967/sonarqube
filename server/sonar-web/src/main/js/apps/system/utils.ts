/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { each, groupBy, memoize, omit, omitBy, pickBy, sortBy } from 'lodash';
import {
  cleanQuery,
  parseAsArray,
  parseAsString,
  RawQuery,
  serializeStringArray
} from '../../helpers/query';
import {
  ClusterSysInfo,
  HealthType,
  NodeInfo,
  SysInfo,
  SysInfoSection,
  SysValueObject,
  SystemUpgrade
} from '../../api/system';
import { formatMeasure } from '../../helpers/measures';

export interface Query {
  expandedCards: string[];
}

export const LOGS_LEVELS = ['INFO', 'DEBUG', 'TRACE'];
export const HA_FIELD = 'High Availability';
export const HEALTH_FIELD = 'Health';
export const HEALTHCAUSES_FIELD = 'Health Causes';
export const PLUGINS_FIELD = 'Plugins';
export const SETTINGS_FIELD = 'Settings';

export function ignoreInfoFields(sysInfoObject: SysValueObject): SysValueObject {
  return omit(sysInfoObject, [
    HEALTH_FIELD,
    HEALTHCAUSES_FIELD,
    'Name',
    PLUGINS_FIELD,
    SETTINGS_FIELD
  ]) as SysValueObject;
}

export function getHealth(sysInfoObject: SysValueObject): HealthType {
  return sysInfoObject[HEALTH_FIELD] as HealthType;
}

export function getHealthCauses(sysInfoObject: SysValueObject): string[] {
  return sysInfoObject[HEALTHCAUSES_FIELD] as string[];
}

export function getLogsLevel(sysInfoObject?: SysValueObject): string {
  if (!sysInfoObject) {
    return LOGS_LEVELS[0];
  }
  if (sysInfoObject['Web Logging'] || sysInfoObject['Compute Engine Logging']) {
    return sortBy(
      [
        getLogsLevel((sysInfoObject as NodeInfo)['Web Logging']),
        getLogsLevel((sysInfoObject as NodeInfo)['Compute Engine Logging'])
      ],
      logLevel => LOGS_LEVELS.indexOf(logLevel)
    )[1];
  }
  if (sysInfoObject['System']) {
    return getLogsLevel((sysInfoObject as SysInfo)['System']);
  }
  return (sysInfoObject['Logs Level'] || LOGS_LEVELS[0]) as string;
}

export function getAppNodes(sysInfoData: ClusterSysInfo): NodeInfo[] {
  return sysInfoData['Application Nodes'];
}

export function getSearchNodes(sysInfoData: ClusterSysInfo): NodeInfo[] {
  return sysInfoData['Search Nodes'];
}

export function isCluster(sysInfoData?: SysInfo): boolean {
  return (
    sysInfoData !== undefined && sysInfoData['System'] && sysInfoData['System'][HA_FIELD] === true
  );
}

export function getServerId(sysInfoData?: SysInfo): string | undefined {
  return sysInfoData && sysInfoData['System']['Server ID'];
}

export function getSystemLogsLevel(sysInfoData?: SysInfo): string {
  const defaultLevel = LOGS_LEVELS[0];
  if (!sysInfoData) {
    return defaultLevel;
  }
  if (isCluster(sysInfoData)) {
    const logLevels = sortBy(
      getAppNodes(sysInfoData as ClusterSysInfo).map(getLogsLevel),
      logLevel => LOGS_LEVELS.indexOf(logLevel)
    );
    return logLevels.length > 0 ? logLevels[logLevels.length - 1] : defaultLevel;
  } else {
    return getLogsLevel(sysInfoData);
  }
}

export function getNodeName(nodeInfo: NodeInfo): string {
  return nodeInfo['Name'];
}

function getSystemData(sysInfoData: SysInfo): SysValueObject {
  const statData: SysValueObject = {};
  const statistics = sysInfoData['Statistics'] as SysValueObject;
  if (statistics) {
    statData['Lines of Code'] = formatMeasure(statistics['ncloc'] as number, 'INT');
  }
  return { ...sysInfoData['System'], ...statData };
}

export function getClusterMainCardSection(sysInfoData: ClusterSysInfo): SysValueObject {
  return {
    ...getSystemData(sysInfoData),
    ...(omit(sysInfoData, [
      'Application Nodes',
      PLUGINS_FIELD,
      'Search Nodes',
      SETTINGS_FIELD,
      'Statistics',
      'System'
    ]) as SysValueObject)
  };
}

export function getStandaloneMainSections(sysInfoData: SysInfo): SysValueObject {
  return {
    ...getSystemData(sysInfoData),
    ...(omitBy(
      sysInfoData,
      (value, key) =>
        value == null ||
        [PLUGINS_FIELD, SETTINGS_FIELD, 'Statistics', 'System'].includes(key) ||
        key.startsWith('Compute Engine') ||
        key.startsWith('Search') ||
        key.startsWith('Web')
    ) as SysValueObject)
  };
}

export function getStandaloneSecondarySections(sysInfoData: SysInfo): SysInfoSection {
  return {
    Web: pickBy(sysInfoData, (_, key) => key.startsWith('Web')) as SysValueObject,
    'Compute Engine': pickBy(sysInfoData, (_, key) =>
      key.startsWith('Compute Engine')
    ) as SysValueObject,
    'Search Engine': pickBy(sysInfoData, (_, key) => key.startsWith('Search')) as SysValueObject
  };
}

export function getFileNameSuffix(suffix?: string) {
  const now = new Date();
  return (
    `${suffix ? suffix + '-' : ''}` +
    `${now.getFullYear()}-${now.getMonth() + 1}-` +
    `${now.getDate()}-${now.getHours()}-${now.getMinutes()}`
  );
}

export function groupSections(sysInfoData: SysValueObject) {
  const mainSection: SysValueObject = {};
  const sections: SysInfoSection = {};
  each(sysInfoData, (item, key) => {
    if (typeof item !== 'object' || item instanceof Array) {
      mainSection[key] = item;
    } else {
      sections[key] = item;
    }
  });
  return { mainSection, sections };
}

export const parseQuery = memoize(
  (urlQuery: RawQuery): Query => ({
    expandedCards: parseAsArray(urlQuery.expand, parseAsString)
  })
);

export const serializeQuery = memoize(
  (query: Query): RawQuery =>
    cleanQuery({
      expand: serializeStringArray(query.expandedCards)
    })
);

export function sortUpgrades(upgrades: SystemUpgrade[]): SystemUpgrade[] {
  return sortBy(upgrades, [
    (upgrade: SystemUpgrade) => -Number(upgrade.version.split('.')[0]),
    (upgrade: SystemUpgrade) => -Number(upgrade.version.split('.')[1] || 0),
    (upgrade: SystemUpgrade) => -Number(upgrade.version.split('.')[2] || 0)
  ]);
}

export function groupUpgrades(upgrades: SystemUpgrade[]): SystemUpgrade[][] {
  const groupedVersions = groupBy(upgrades, upgrade => upgrade.version.split('.')[0]);
  const sortedMajor = sortBy(Object.keys(groupedVersions), key => -Number(key));
  return sortedMajor.map(key => groupedVersions[key]);
}
