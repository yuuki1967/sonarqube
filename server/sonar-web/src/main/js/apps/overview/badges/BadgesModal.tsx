/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import BadgeButton from './BadgeButton';
import BadgeParams from './BadgeParams';
import { BadgeType, BadgeOptions, getBadgeUrl, getBadgeSnippet } from './utils';
import CodeSnippet from '../../../components/common/CodeSnippet';
import Modal from '../../../components/controls/Modal';
import { getBranchLikeQuery } from '../../../helpers/branches';
import { translate } from '../../../helpers/l10n';
import { Button, ResetButtonLink } from '../../../components/ui/buttons';
import { isSonarCloud } from '../../../helpers/system';
import './styles.css';

interface Props {
  branchLike?: T.BranchLike;
  metrics: { [key: string]: T.Metric };
  project: string;
  qualifier: string;
}

interface State {
  open: boolean;
  selectedType: BadgeType;
  badgeOptions: BadgeOptions;
}

export default class BadgesModal extends React.PureComponent<Props, State> {
  state: State = {
    open: false,
    selectedType: BadgeType.measure,
    badgeOptions: { color: 'white', metric: 'alert_status' }
  };

  handleClose = () => {
    this.setState({ open: false });
  };

  handleOpen = () => {
    this.setState({ open: true });
  };

  handleSelectBadge = (selectedType: BadgeType) => {
    this.setState({ selectedType });
  };

  handleUpdateOptions = (options: Partial<BadgeOptions>) => {
    this.setState(state => ({ badgeOptions: { ...state.badgeOptions, ...options } }));
  };

  render() {
    const { branchLike, project, qualifier } = this.props;
    const { selectedType, badgeOptions } = this.state;
    const header = translate('overview.badges.title');
    const fullBadgeOptions = { project, ...badgeOptions, ...getBranchLikeQuery(branchLike) };
    const badges = isSonarCloud()
      ? [BadgeType.measure, BadgeType.qualityGate, BadgeType.marketing]
      : [BadgeType.measure, BadgeType.qualityGate];
    return (
      <div className="overview-meta-card">
        <Button className="js-project-badges" onClick={this.handleOpen}>
          {translate('overview.badges.get_badge', qualifier)}
        </Button>
        {this.state.open && (
          <Modal contentLabel={header} onRequestClose={this.handleClose}>
            <header className="modal-head">
              <h2>{header}</h2>
            </header>
            <div className="modal-body">
              <p className="huge-spacer-bottom">
                {translate('overview.badges.description', qualifier)}
              </p>
              <div className="badges-list spacer-bottom">
                {badges.map(type => (
                  <BadgeButton
                    key={type}
                    onClick={this.handleSelectBadge}
                    selected={type === selectedType}
                    type={type}
                    url={getBadgeUrl(type, fullBadgeOptions)}
                  />
                ))}
              </div>
              <p className="text-center note huge-spacer-bottom">
                {translate('overview.badges', selectedType, 'description', qualifier)}
              </p>
              <BadgeParams
                className="big-spacer-bottom"
                metrics={this.props.metrics}
                options={badgeOptions}
                type={selectedType}
                updateOptions={this.handleUpdateOptions}
              />
              <CodeSnippet
                isOneLine={true}
                snippet={getBadgeSnippet(selectedType, fullBadgeOptions)}
              />
            </div>
            <footer className="modal-foot">
              <ResetButtonLink className="js-modal-close" onClick={this.handleClose}>
                {translate('close')}
              </ResetButtonLink>
            </footer>
          </Modal>
        )}
      </div>
    );
  }
}
