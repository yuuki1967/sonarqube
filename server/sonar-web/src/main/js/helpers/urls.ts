/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { stringify } from 'querystring';
import { omitBy, isNil } from 'lodash';
import {
  isShortLivingBranch,
  isPullRequest,
  isLongLivingBranch,
  getBranchLikeQuery
} from './branches';
import { getProfilePath } from '../apps/quality-profiles/utils';

interface Query {
  [x: string]: string | undefined;
}

export interface Location {
  pathname: string;
  query?: Query;
}

export function getBaseUrl(): string {
  return (window as any).baseUrl;
}

export function getHostUrl(): string {
  return window.location.origin + getBaseUrl();
}

export function getPathUrlAsString(path: Location, internal = true): string {
  return `${internal ? getBaseUrl() : getHostUrl()}${path.pathname}?${stringify(
    omitBy(path.query, isNil)
  )}`;
}

export function getProjectUrl(project: string, branch?: string): Location {
  return { pathname: '/dashboard', query: { id: project, branch } };
}

export function getPortfolioUrl(key: string): Location {
  return { pathname: '/portfolio', query: { id: key } };
}

export function getPortfolioAdminUrl(key: string, qualifier: string) {
  return { pathname: '/project/admin/extension/governance/console', query: { id: key, qualifier } };
}

export function getComponentBackgroundTaskUrl(componentKey: string, status?: string): Location {
  return { pathname: '/project/background_tasks', query: { id: componentKey, status } };
}

export function getBranchLikeUrl(project: string, branchLike?: T.BranchLike): Location {
  if (isPullRequest(branchLike)) {
    return getPullRequestUrl(project, branchLike.key);
  } else if (isShortLivingBranch(branchLike)) {
    return getShortLivingBranchUrl(project, branchLike.name);
  } else if (isLongLivingBranch(branchLike)) {
    return getLongLivingBranchUrl(project, branchLike.name);
  } else {
    return getProjectUrl(project);
  }
}

export function getLongLivingBranchUrl(project: string, branch: string): Location {
  return { pathname: '/dashboard', query: { branch, id: project } };
}

export function getShortLivingBranchUrl(project: string, branch: string): Location {
  return { pathname: '/dashboard', query: { branch, id: project } };
}

export function getPullRequestUrl(project: string, pullRequest: string): Location {
  return { pathname: '/dashboard', query: { id: project, pullRequest } };
}

/**
 * Generate URL for a global issues page
 */
export function getIssuesUrl(query: Query, organization?: string): Location {
  const pathname = organization ? `/organizations/${organization}/issues` : '/issues';
  return { pathname, query };
}

/**
 * Generate URL for a component's issues page
 */
export function getComponentIssuesUrl(componentKey: string, query?: Query): Location {
  return { pathname: '/project/issues', query: { ...(query || {}), id: componentKey } };
}

/**
 * Generate URL for a component's drilldown page
 */
export function getComponentDrilldownUrl(options: {
  componentKey: string;
  metric: string;
  branchLike?: T.BranchLike;
  selectionKey?: string;
  treemapView?: boolean;
  listView?: boolean;
}): Location {
  const { componentKey, metric, branchLike, selectionKey, treemapView, listView } = options;
  const query: Query = { id: componentKey, metric, ...getBranchLikeQuery(branchLike) };
  if (treemapView) {
    query.view = 'treemap';
  }
  if (listView) {
    query.view = 'list';
  }
  if (selectionKey) {
    query.selected = selectionKey;
  }
  return { pathname: '/component_measures', query };
}

export function getComponentDrilldownUrlWithSelection(
  componentKey: string,
  selectionKey: string,
  metric: string,
  branchLike?: T.BranchLike
): Location {
  return getComponentDrilldownUrl({ componentKey, selectionKey, metric, branchLike });
}

export function getMeasureTreemapUrl(componentKey: string, metric: string) {
  return getComponentDrilldownUrl({ componentKey, metric, treemapView: true });
}

export function getActivityUrl(component: string, branchLike?: T.BranchLike) {
  return {
    pathname: '/project/activity',
    query: { id: component, ...getBranchLikeQuery(branchLike) }
  };
}

/**
 * Generate URL for a component's measure history
 */
export function getMeasureHistoryUrl(component: string, metric: string, branchLike?: T.BranchLike) {
  return {
    pathname: '/project/activity',
    query: {
      id: component,
      graph: 'custom',
      custom_metrics: metric,
      ...getBranchLikeQuery(branchLike)
    }
  };
}

/**
 * Generate URL for a component's permissions page
 */
export function getComponentPermissionsUrl(componentKey: string): Location {
  return { pathname: '/project_roles', query: { id: componentKey } };
}

/**
 * Generate URL for a quality profile
 */
export function getQualityProfileUrl(
  name: string,
  language: string,
  organization?: string | null
): Location {
  return getProfilePath(name, language, organization);
}

export function getQualityGateUrl(key: string, organization?: string | null): Location {
  return {
    pathname: getQualityGatesUrl(organization).pathname + '/show/' + encodeURIComponent(key)
  };
}

export function getQualityGatesUrl(organization?: string | null): Location {
  return {
    pathname:
      (organization ? '/organizations/' + encodeURIComponent(organization) : '') + '/quality_gates'
  };
}

/**
 * Generate URL for the rules page
 */
export function getRulesUrl(query: Query, organization: string | null | undefined): Location {
  const pathname = organization ? `/organizations/${organization}/rules` : '/coding_rules';
  return { pathname, query };
}

/**
 * Generate URL for the rules page filtering only active deprecated rules
 */
export function getDeprecatedActiveRulesUrl(
  query: Query = {},
  organization: string | null | undefined
): Location {
  const baseQuery = { activation: 'true', statuses: 'DEPRECATED' };
  return getRulesUrl({ ...query, ...baseQuery }, organization);
}

export function getRuleUrl(rule: string, organization: string | undefined) {
  return getRulesUrl({ open: rule, rule_key: rule }, organization);
}

export function getMarkdownHelpUrl(): string {
  return getBaseUrl() + '/markdown/help';
}

export function getCodeUrl(
  project: string,
  branchLike?: T.BranchLike,
  selected?: string,
  line?: number
) {
  return {
    pathname: '/code',
    query: { id: project, ...getBranchLikeQuery(branchLike), selected, line }
  };
}

export function getOrganizationUrl(organization: string) {
  return `/organizations/${organization}`;
}

export function getHomePageUrl(homepage: T.HomePage) {
  switch (homepage.type) {
    case 'APPLICATION':
      return homepage.branch
        ? getProjectUrl(homepage.component, homepage.branch)
        : getProjectUrl(homepage.component);
    case 'PROJECT':
      return homepage.branch
        ? getLongLivingBranchUrl(homepage.component, homepage.branch)
        : getProjectUrl(homepage.component);
    case 'ORGANIZATION':
      return getOrganizationUrl(homepage.organization);
    case 'PORTFOLIO':
      return getPortfolioUrl(homepage.component);
    case 'PORTFOLIOS':
      return '/portfolios';
    case 'MY_PROJECTS':
      return '/projects';
    case 'ISSUES':
    case 'MY_ISSUES':
      return { pathname: '/issues', query: { resolved: 'false' } };
  }

  // should never happen, but just in case...
  return '/projects';
}

export function getReturnUrl(location: { hash?: string; query?: { return_to?: string } }) {
  const returnTo = location.query && location.query['return_to'];
  if (isRelativeUrl(returnTo)) {
    return returnTo + (location.hash ? location.hash : '');
  }
  return getBaseUrl() + '/';
}

export function isRelativeUrl(url?: string): boolean {
  const regex = new RegExp(/^\/[^/\\]/);
  return Boolean(url && regex.test(url));
}
