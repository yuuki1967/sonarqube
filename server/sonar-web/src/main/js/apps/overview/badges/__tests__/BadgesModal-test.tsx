/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import BadgesModal from '../BadgesModal';
import { click } from '../../../../helpers/testUtils';
import { isSonarCloud } from '../../../../helpers/system';
import { Location } from '../../../../helpers/urls';

jest.mock('../../../../helpers/urls', () => ({
  getHostUrl: () => 'host',
  getProjectUrl: () => ({ pathname: '/dashboard' } as Location),
  getPathUrlAsString: (l: Location) => l.pathname
}));
jest.mock('../../../../helpers/system', () => ({ isSonarCloud: jest.fn() }));

const shortBranch: T.ShortLivingBranch = {
  isMain: false,
  mergeBranch: '',
  name: 'branch-6.6',
  type: 'SHORT'
};

it('should display the modal after click on sonarcloud', () => {
  (isSonarCloud as jest.Mock).mockImplementation(() => true);
  const wrapper = shallow(
    <BadgesModal branchLike={shortBranch} metrics={{}} project="foo" qualifier="TRK" />
  );
  expect(wrapper).toMatchSnapshot();
  click(wrapper.find('Button'));
  expect(wrapper.find('Modal')).toMatchSnapshot();
});

it('should display the modal after click on sonarqube', () => {
  (isSonarCloud as jest.Mock).mockImplementation(() => false);
  const wrapper = shallow(
    <BadgesModal branchLike={shortBranch} metrics={{}} project="foo" qualifier="TRK" />
  );
  expect(wrapper).toMatchSnapshot();
  click(wrapper.find('Button'));
  expect(wrapper.find('Modal')).toMatchSnapshot();
});
