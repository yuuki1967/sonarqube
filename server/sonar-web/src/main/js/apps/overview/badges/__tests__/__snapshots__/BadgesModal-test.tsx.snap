// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the modal after click on sonarcloud 1`] = `
<div
  className="overview-meta-card"
>
  <Button
    className="js-project-badges"
    onClick={[Function]}
  >
    overview.badges.get_badge.TRK
  </Button>
</div>
`;

exports[`should display the modal after click on sonarcloud 2`] = `
<Modal
  contentLabel="overview.badges.title"
  onRequestClose={[Function]}
>
  <header
    className="modal-head"
  >
    <h2>
      overview.badges.title
    </h2>
  </header>
  <div
    className="modal-body"
  >
    <p
      className="huge-spacer-bottom"
    >
      overview.badges.description.TRK
    </p>
    <div
      className="badges-list spacer-bottom"
    >
      <BadgeButton
        key="measure"
        onClick={[Function]}
        selected={true}
        type="measure"
        url="host/api/project_badges/measure?branch=branch-6.6&project=foo&metric=alert_status"
      />
      <BadgeButton
        key="quality_gate"
        onClick={[Function]}
        selected={false}
        type="quality_gate"
        url="host/api/project_badges/quality_gate?branch=branch-6.6&project=foo"
      />
      <BadgeButton
        key="marketing"
        onClick={[Function]}
        selected={false}
        type="marketing"
        url="host/images/project_badges/sonarcloud-white.svg"
      />
    </div>
    <p
      className="text-center note huge-spacer-bottom"
    >
      overview.badges.measure.description.TRK
    </p>
    <BadgeParams
      className="big-spacer-bottom"
      metrics={Object {}}
      options={
        Object {
          "color": "white",
          "metric": "alert_status",
        }
      }
      type="measure"
      updateOptions={[Function]}
    />
    <CodeSnippet
      isOneLine={true}
      snippet="[![alert_status](host/api/project_badges/measure?branch=branch-6.6&project=foo&metric=alert_status)](/dashboard)"
    />
  </div>
  <footer
    className="modal-foot"
  >
    <ResetButtonLink
      className="js-modal-close"
      onClick={[Function]}
    >
      close
    </ResetButtonLink>
  </footer>
</Modal>
`;

exports[`should display the modal after click on sonarqube 1`] = `
<div
  className="overview-meta-card"
>
  <Button
    className="js-project-badges"
    onClick={[Function]}
  >
    overview.badges.get_badge.TRK
  </Button>
</div>
`;

exports[`should display the modal after click on sonarqube 2`] = `
<Modal
  contentLabel="overview.badges.title"
  onRequestClose={[Function]}
>
  <header
    className="modal-head"
  >
    <h2>
      overview.badges.title
    </h2>
  </header>
  <div
    className="modal-body"
  >
    <p
      className="huge-spacer-bottom"
    >
      overview.badges.description.TRK
    </p>
    <div
      className="badges-list spacer-bottom"
    >
      <BadgeButton
        key="measure"
        onClick={[Function]}
        selected={true}
        type="measure"
        url="host/api/project_badges/measure?branch=branch-6.6&project=foo&metric=alert_status"
      />
      <BadgeButton
        key="quality_gate"
        onClick={[Function]}
        selected={false}
        type="quality_gate"
        url="host/api/project_badges/quality_gate?branch=branch-6.6&project=foo"
      />
    </div>
    <p
      className="text-center note huge-spacer-bottom"
    >
      overview.badges.measure.description.TRK
    </p>
    <BadgeParams
      className="big-spacer-bottom"
      metrics={Object {}}
      options={
        Object {
          "color": "white",
          "metric": "alert_status",
        }
      }
      type="measure"
      updateOptions={[Function]}
    />
    <CodeSnippet
      isOneLine={true}
      snippet="[![alert_status](host/api/project_badges/measure?branch=branch-6.6&project=foo&metric=alert_status)](/dashboard)"
    />
  </div>
  <footer
    className="modal-foot"
  >
    <ResetButtonLink
      className="js-modal-close"
      onClick={[Function]}
    >
      close
    </ResetButtonLink>
  </footer>
</Modal>
`;
